### PR TITLE
grt: fix congestion issues

### PR DESCRIPTION
**Pull request by eder-matheus**
- Added check for huge usage values in 2D edges
- Added limitation to the number of total overflow increases during overflow iterations
- Overflow in the final routing is allowed now
- Four unit tests for congestion, including a case that tries to replicate the issue in [OR-1032](https://openroad.atlassian.net/browse/OR-1032)

### CODE REVIEW

File: src/grt/src/fastroute/src/utility.cpp [hunk 1]
```diff
@@ -1334,6 +1334,37 @@ void checkUsage()
...
+  int max_v_edge_usage = 40 * vCapacity;
+
+  // check horizontal edges
+  for (int i = 0; i < yGrid; i++) {
+    for (int j = 0; j < xGrid - 1; j++) {
+      int grid = i * (xGrid - 1) + j;
+      if (h_edges[grid].usage >= max_h_edge_usage) {
+        logger->error(GRT,
+                      228,
+                      "Horizontal edge exceeds the maximum usage allowed.");
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:1350]: Horizontal edge usage exceeds the maximum allowed

File: src/grt/src/fastroute/src/utility.cpp [hunk 2]
```diff
@@ -1334,6 +1334,37 @@ void checkUsage()
   }
 }
 
+void check2DEdgesUsage()
+{
+  // check maze.cpp:1105-1106 for these values
+  int max_h_edge_usage = 40 * hCapacity;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/utility.cpp:1340]: Make 40 into a named constant that is shared by both pieces of code.

**Review by maliberty — CHANGES_REQUESTED** [review]
Why are we allowing routing to end with overflow by default?  We will be passing an infeasible solution onto detailed routing which will likely fail too.

**Comment by eder-matheus** [comment]
> Why are we allowing routing to end with overflow by default? We will be passing an infeasible solution onto detailed routing which will likely fail too.

You're right, I think I made confusion here. I will update my code to allow overflow during the routing process (removing the `allowOverflow` flag from the core code), but I'll keep the check after the routing is complete. Therefore, I'll restore the `-allo_overflow` option (and probably change it to `-allow_congestion`.

**Comment by eder-matheus** [comment]
@maliberty any further comments/requests?

**Review by maliberty — APPROVED** [review]

**Pull request merged**